routing/unified_policies.go

```diff
@@ -0,0 +1,263 @@
+		maxBandwidth lnwire.MilliSatoshi
+	)
+
+	for _, edge := range u.edges {
+		// Check valid amount range for the channel.
+		if !edge.amtInRange(amt) {
+			continue
+		}
+
+		// For local channels, there is no fee to pay or an extra time
+		// lock. We only consider the currently available bandwidth for
+		// channel selection. The disabled flag is ignored for local
+		// channels.
+
+		// Retrieve bandwidth for this local channel. If not
+		// available, assume this channel has enough bandwidth.
+		//
+		// TODO(joostjager): Possibly change to skipping this
+		// channel. The bandwidth hint is expected to be
+		// available.
+		bandwidth, ok := bandwidthHints[edge.policy.ChannelID]
+		if !ok {
+			bandwidth = lnwire.MaxMilliSatoshi
+		}
+
+		// Skip channels that can't carry the payment.
+		if amt > bandwidth {
+			continue
+		}
+
+		// We pick the local channel with the highest available
+		// bandwidth, to maximize the success probability. It
+		// can be that the channel state changes between
+		// querying the bandwidth hints and sending out the
+		// htlc.
+		if bandwidth < maxBandwidth {
+			continue
+		}
+		maxBandwidth = bandwidth
+
+		// Update best policy.
+		bestPolicy = edge.policy
+	}
+
+	return bestPolicy
+}
+
+// getPolicyNetwork returns the optimal policy to use for this connection given
+// a specific amount to send. The goal is to return a policy that maximizes the
+// probability of a successful forward in a non-strict forwarding context.
+func (u *unifiedPolicy) getPolicyNetwork(
+	amt lnwire.MilliSatoshi) *channeldb.ChannelEdgePolicy {
+
+	var (
+		bestPolicy  *channeldb.ChannelEdgePolicy
+		maxFee      lnwire.MilliSatoshi
+		maxTimelock uint16
+	)
+
+	for _, edge := range u.edges {
+		// Check valid amount range for the channel.
+		if !edge.amtInRange(amt) {
+			continue
+		}
+
+		// For network channels, skip the disabled ones.
+		edgeFlags := edge.policy.ChannelFlags
+		isDisabled := edgeFlags&lnwire.ChanUpdateDisabled != 0
+		if isDisabled {
+			continue
+		}
+
+		// Track the maximum time lock of all channels that are
+		// candidate for non-strict forwarding at the routing node.
+		if edge.policy.TimeLockDelta > maxTimelock {
+			maxTimelock = edge.policy.TimeLockDelta
+		}
+
+		// Use the policy that results in the highest fee for this
+		// specific amount.
+		fee := edge.policy.ComputeFee(amt)
+		if fee < maxFee {
+			continue
+		}
+		maxFee = fee
+
+		bestPolicy = edge.policy
+	}
+
+	// Return early if no channel matches.
+	if bestPolicy == nil {
+		return nil
+	}
+
+	// We have already picked the highest fee that could be required for
+	// non-strict forwarding. To also cover the case where a lower fee
+	// channel requires a longer time lock, we modify the policy by setting
+	// the maximum encountered time lock. Note that this results in a
+	// synthetic policy that is not actually present on the routing node.
+	modifiedPolicy := *bestPolicy
+	modifiedPolicy.TimeLockDelta = maxTimelock
```

Comment:
If I understand correctly, this is done so that there isn't an additional round trip in the payment process in the event that the channel yielding the highest fee does not have sufficient bandwidth to carry the payment but the highest timelock one does?

Reply:
Yes indeed. We try to maximize the chance that we get forwarded. Because we penalize pair-wise, there won't be a second chance for this node pair. But this all really only exists for nodes that have distinct policies for channels to the same peer. It seems unlikely it is configured like that intentionally. Currently it does happen on the network after upgrading to a `lnd` version that changed the default time lock delta. The new delta will be applied to newly opened channels.

Reply:
Nice, I think including this insight in the comment would be helpful.

Reply:
Included